src/component/project/ProjectEnvironment/ProjectEnvironment.tsx

```diff
@@ -0,0 +1,141 @@
+import { useContext, useState } from 'react';
+import { useParams } from 'react-router-dom';
+import ConditionallyRender from '../../common/ConditionallyRender';
+import { useStyles } from './ProjectEnvironment.styles';
+
+import useLoading from '../../../hooks/useLoading';
+import PageContent from '../../common/PageContent';
+import AccessContext from '../../../contexts/AccessContext';
+import HeaderTitle from '../../common/HeaderTitle';
+import { UPDATE_PROJECT } from '../../AccessProvider/permissions';
+
+import ApiError from '../../common/ApiError/ApiError';
+import useToast from '../../../hooks/useToast';
+import useUiConfig from '../../../hooks/api/getters/useUiConfig/useUiConfig';
+import useEnvironments from '../../../hooks/api/getters/useEnvironments/useEnvironments';
+import useProject from '../../../hooks/api/getters/useProject/useProject';
+import { FormControlLabel, FormGroup, Switch } from '@material-ui/core';
+import useProjectApi from '../../../hooks/api/actions/useProjectApi/useProjectApi';
+import EnvironmentDisableConfirm from './EnvironmentDisableConfirm/EnvironmentDisableConfirm';
+
+export interface ProjectEnvironment {
+    name: string;
+    enabled: boolean;
+}
+
+const ProjectListNew = () => {
+    const { id } = useParams<{id: string}>();
+    const { hasAccess } = useContext(AccessContext);
+
+    // api state
+    const { toast } = useToast();
+    const { uiConfig } = useUiConfig();
+    const { environments, loading, error } = useEnvironments();
+    const { project, refetch } = useProject(id);
+    const { removeEnvironmentFromProject, addEnvironmentToProject } = useProjectApi();
+    
+    // local state
+    const [selectedEnv, setSelectedEnv] = useState<ProjectEnvironment>();
+    const [confirmName, setConfirmName] = useState('');
+    const ref = useLoading(loading);
+    const styles = useStyles();
+
+
+    const renderError = () => {
+        return (
+            <ApiError
+                onClick={refetch}
+                className={styles.apiError}
+                text="Error fetching environments"
+            />
+        );
+    };
+
+    const toggleEnv = async (env: ProjectEnvironment) => {
+        if(env.enabled) {
+            // await removeEnvironmentFromProject(id, env.name);
+            setSelectedEnv(env);
+
+        } else {
+            await addEnvironmentToProject(id, env.name);
+        }
+        refetch();
+    }
+
+    const handleDisableEnvironment = async () => {
+        if(selectedEnv && confirmName===selectedEnv.name) {
+            await removeEnvironmentFromProject(id, selectedEnv.name);
```

Comment:
Same here, should be wrapped in a try catch. Alternatively you could set propagateErrors to false and let the error be handled by the useApi parent hook. This will handle the most common usecases and set the error in the errors object exposed from the hook.

Reply:
thanks for explaining this!

Reply:
No worries! I added propagation later because sometimes you want to access the error message if you are setting a toast or something similar.

Reply:
in useApi.ts you can see the defaults: 

It checks for bad request, not found, unauthorized and forbidden. 

The keys are set as "unauthorized", "notFound", "badRequest", "forbidden". Now that I think about it, it is a bit convoluted that you have to check for the keys manually, but one could argue that you would like to perform different UI actions based on error you set. We could implement a general component that would resolve it for you. 

Lately, I have found the propagation easier, because in the component I generally know what feedback I would like to show.

Reply:
I will go for the propagation for now.